### PR TITLE
Runtime-reconfigurable cache sizes

### LINKED ISSUE #51085: Dynamically update `mark_cache_size`, `index_mark_cache_size` and `mmap_cache_size` from config
Now, `mark_cache_size`, `index_mark_cache_size` and `mmap_cache_size` are set at server startup and never changes.

https://github.com/ClickHouse/ClickHouse/blob/1cb7ba4895e0666eeba8654a45ec8c124b904864/programs/server/Server.cpp#L1459-L1466

We can support that settings into `ConfigReloader` as well. Caches should support dynamic changes in size.
In case if cache uses more memory than changed limit, some entries should be ejected instantly.
At least for LRU it should be easy: just call `removeOverflow`
https://github.com/ClickHouse/ClickHouse/blob/ab6ac6beb19ae23d0676f7626c6fa9b72353de3e/src/Common/LRUCachePolicy.h#L128

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes #51085 

Note: Only https://github.com/ClickHouse/ClickHouse/pull/51446/commits/96c8b250ce39770df0cc296a880d327d4d1b8136 contains relevant changes, the other commit has sort-of-unrelated some cleanups.

Only
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The sizes of the (index) uncompressed/mark, mmap and query caches can now be configured dynamically at runtime.

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit 6203b1c0c674a1603116bb39974245b96df062d4 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51446/6203b1c0c674a1603116bb39974245b96df062d4/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by rschu1ze** [comment]
The test failures are caused by a new sanity check. I'll go through all usages of the caches and simplify.

File: src/Common/SLRUCachePolicy.h [hunk 1]
```diff
@@ -54,7 +54,22 @@ class SLRUCachePolicy : public ICachePolicy<Key, Mapped, HashFunction, WeightFun
...
+    void setMaxCount(size_t max_count_, std::lock_guard<std::mutex> & /* cache_lock */) override
+    {
+        max_count = max_count_;
+        removeOverflow(protected_queue, max_protected_size, current_protected_size, /*is_protected=*/true);
+        removeOverflow(probationary_queue, max_size_in_bytes, current_size_in_bytes, /*is_protected=*/false);
+    }
+
+    void setMaxSize(size_t max_size_in_bytes_, std::lock_guard<std::mutex> & /* cache_lock */) override
+    {
+        max_protected_size = static_cast<size_t>(max_size_in_bytes_ * std::min(1.0, size_ratio));
```
Comment (KochetovNicolai) [inline src/Common/SLRUCachePolicy.h:66]: Used in ctor, maybe eject into function.

Does it make sense to allow updating size_ratio dynamically?

**Review by KochetovNicolai — APPROVED** [review]

File: tests/integration/test_runtime_configurable_cache_size/test.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    node.query("INSERT INTO test1 VALUES ('abc') ('def') ('ghi')")
+    node.query("SELECT * FROM test1 WHERE val = 'def'")  # cache 1st mark
+
+    node.query("CREATE TABLE test2 (val String) ENGINE=MergeTree ORDER BY val")
+    node.query("INSERT INTO test2 VALUES ('abc') ('def') ('ghi')")
+    node.query("SELECT * FROM test2 WHERE val = 'def'")  # cache 2nd mark
+
+    # Result checking is based on asynchronous metrics. These are calculated by default every 1.0 sec, and this is also the
+    # smallest possible value. Found no statement to force-recalculate them, therefore waaaaait...
+    time.sleep(2.0)
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply tests/integration/test_runtime_configurable_cache_size/test.py:46]: I also don't like the sleeps, and thought about already about synchronous metrics. I'll check separately how the latency requirements for synchronous metrics are (we'd need to acquire locks to fetch cache statistics ...).

**Comment by rschu1ze** [comment]
Integration test complains that mark cache entries are slightly smaller in CI than locally (i.e. in expected results). Let me disable this particular test temporarily and merge (someone is waiting for this PR), I'll fix stuff separately.

```

=================================== FAILURES ===================================
_________________ test_mark_cache_size_is_runtime_configurable _________________
[gw0] linux -- Python 3.10.12 /usr/bin/python3

start_cluster = <helpers.cluster.ClickHouseCluster object at 0x7f4e37d134f0>

    def test_mark_cache_size_is_runtime_configurable(start_cluster):
        # the initial config specifies the mark cache size as 496 bytes, just enough to hold two marks
        node.query("SYSTEM DROP MARK CACHE")
    
        node.query("CREATE TABLE test1 (val String) ENGINE=MergeTree ORDER BY val")
        node.query("INSERT INTO test1 VALUES ('abc') ('def') ('ghi')")
        node.query("SELECT * FROM test1 WHERE val = 'def'")  # cache 1st mark
    
        node.query("CREATE TABLE test2 (val String) ENGINE=MergeTree ORDER BY val")
        node.query("INSERT INTO test2 VALUES ('abc') ('def') ('ghi')")
        node.query("SELECT * FROM test2 WHERE val = 'def'")  # cache 2nd mark
    
        # Result checking is based on asynchronous metrics. These are calculated by default every 1.0 sec, and this is also the
        # smallest possible value. Found no statement to force-recalculate them, therefore waaaaait...
        time.sleep(2.0)
        res = node.query(
            "SELECT value FROM system.asynchronous_metrics WHERE metric LIKE 'MarkCacheFiles'"
        )
        assert res == "2\n"
        res = node.query(
            "SELECT value FROM system.asynchronous_metrics WHERE metric LIKE 'MarkCacheBytes'"
        )
>       assert res == "496\n"
E       AssertionError: assert '464\n' == '496\n'
E         - 496
E         ?  -
E         + 464
E         ?   +
```

**Pull request merged**